Federated Server API

dgaubert opened the pull request.

gonzaloriestra commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,61 @@
+
+module Carto
+  class FederatedTablesService
+    def initialize(user, page, per_page, order, direction, offset)
```
I prefer to use keyword arguments when the method has several: https://thoughtbot.com/blog/ruby-2-keyword-arguments. That way it's easier to understand.

gonzaloriestra commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,52 @@
+      class FederatedTablesController < Carto::Api::Public::ApplicationController
+        include Carto::Api::PagedSearcher
+        include Carto::ControllerHelper
+        extend Carto::DefaultRescueFroms
```
You also have to add `setup_default_rescues` to use this.

dgaubert commented:
Yeah! WIP totally, thanks for the tips. Going to implement them.

dgaubert commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,61 @@
+      @order = order
+      @direction = direction
+      @offset = offset
+      @superuser_db_connection = @user.in_database(as: :superuser)
```
Yes, it is. I left there because I'm gonna need it, it was a reminder. :)

dgaubert replied:
Commenting that line for the sake of cleanliness.

dgaubert commented:
Hey @gonzaloriestra!

It's still in progress but I finished the RoR endpoints implementation. Before using the final queries to handle Federated Server & Remote Tables, I'd like you to review it.

Thanks!

dgaubert commented:
> And one last suggestion for you: https://hire.withgoogle.com/public/jobs/cartodbcom/view/P_AAAAAAHAACdKWm_bbm55gX

:)

dgaubert commented on config/routes.rb:
```diff
@@ -597,9 +597,23 @@
+      post 'federated_servers/:federated_server_name/remote_schemas/:remote_schema_name/remote_tables', to: 'federated_tables#register_remote_table', as: :api_v4_federated_servers_register_table
+      get 'federated_servers/:federated_server_name/remote_schemas/:remote_schema_name/remote_tables/:remote_table_name', to: 'federated_tables#show_remote_table', as: :api_v4_federated_servers_get_table
+      put 'federated_servers/:federated_server_name/remote_schemas/:remote_schema_name/remote_tables/:remote_table_name', to: 'federated_tables#update_remote_table', as: :api_v4_federated_servers_update_table
+      delete 'federated_servers/:federated_server_name/remote_schemas/:remote_schema_name/remote_tables/:remote_table_name', to: 'federated_tables#unregister_remote_table', as: :api_v4_federated_servers_unregister_table
```
I'm gonna leave this for a while as I need to integrate the new extension functions first.

Thanks for the tips.

dgaubert replied:
@gonzaloriestra, It could be possible to leave the current controller structure as it is?. The main idea is to release a Beta version so, in following iterations, we are going to validate the API, entry points, parameters and so on. This is going to change and keep it in a single file is more convenient for us by now.

If you agree, I'll open an issue to achieve this refactor as soon as we get a more stable API.

gonzaloriestra replied:
:+1:

alrocar commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,208 @@
+        def register_federated_server
+          federated_server = @service.register_server(@federated_server_attributes)
+          response.headers['Content-Location'] = "#{request.path}/#{federated_server[:federated_server_name]}"
+          render_jsonp({}, 201)
```
to be consistent with other APIs, when returning a 201 we should return a JSON instance of the object created.

Adding the comment here, but could be applied to the rest of the methods in this controller returning a 201.

dgaubert replied:
I wanted to be consistent with the REST specification (https://tools.ietf.org/html/rfc7231#section-4.3.3). I don't mind to change it but we might make a decision here: be consistent with the project or the standard.

alrocar replied:
Good point 🙀 

if you ask me, sticking to the standard is what would be giving us consistency among our APIs. Until now, it seems that's not always been the case, so difficult decision to make.

Summoning @gonzaloriestra to hear a third opinion.

gonzaloriestra replied:
Nobody follows a 100% RESTful implementation (if that's even possible), and I think it's more important to be consistent in the project. As we already have several public APIs working that way, it wouldn't be easy to change it everywhere, so I'd keep the behavior.

Also, this is the default response in Rails scaffolds, and returning more information than expected is not a big deal usually.

dgaubert replied:
Ok, fair enough. I'm going to keep the project's behavior. Thanks!

alrocar replied:
👍

dgaubert replied:
Done

alrocar commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,208 @@
+          raise Carto::CartoError.new("Missing 'port' attribute", 422) unless params[:port].present?
+          raise Carto::CartoError.new("Missing 'username' attribute", 422) unless params[:username].present?
+          raise Carto::CartoError.new("Missing 'password' attribute", 422) unless params[:password].present?
+        end
```
We have a `ensure_required_params` method in the `ApplicationController` that could be handy for these validations.

gonzaloriestra replied:
Another way: https://edgeapi.rubyonrails.org/classes/ActionController/StrongParameters.html

dgaubert replied:
using `ensure_required_params ` returns status code 400. In other places we use 422 instead. What should we use?

dgaubert replied:
done

alrocar commented on spec/requests/carto/api/public/federated_tables_controller_spec.rb:
```diff
@@ -0,0 +1,719 @@
+require 'spec_helper_min'
```
🔝 tests

Algunenano commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,208 @@
+        end
+
+        def check_federated_server_attributes
+          if request.post?
```
Database name and port should be optional, they have well established defaults in PG.

dgaubert commented:
retest this please

dgaubert commented:
retest this please

dgaubert commented:
Locally test are failing because of the following:

```
$ bundle exec rspec spec/requests/carto/api/public/federated_tables_controller_spec.rb:326
```

```
ERROR:  Not enough permissions to access the server \"fs_007_from_user00000001_to_user00000003\"
CONTEXT:  PL/pgSQL function __cdb_fs_create_schema(name,text) line 15 at RAISE
PL/pgSQL function __cdb_fs_list_foreign_schemas_pg(name) line 17 during statement block local variable initialization
PL/pgSQL function cdb_federated_server_list_remote_schemas(text) line 8 at RETURN QUERY
```
To get the response error:
```
$ git diff
diff --git a/spec/requests/carto/api/public/federated_tables_controller_spec.rb b/spec/requests/carto/api/public/federated_tables_controller_spec.rb
index f06642b..170d28e 100644
--- a/spec/requests/carto/api/public/federated_tables_controller_spec.rb
+++ b/spec/requests/carto/api/public/federated_tables_controller_spec.rb
@@ -326,6 +326,7 @@ describe Carto::Api::Public::FederatedTablesController do
     it 'returns 200 with the remote schemas list' do
       params_list_schemas = { federated_server_name: @federated_server_name, api_key: @user1.api_key, page: 1, per_page: 10 }
       get_json api_v4_federated_servers_list_schemas_url(params_list_schemas) do |response|
+        puts response.body
         expect(response.status).to eq(200)
         expect(response.body[:total] > 0)
         found = response.body[:result].select {|schema| schema[:remote_schema_name] == 'public'}.first
```

When I use the superuser connection:

```
$ git diff
diff --git a/app/services/carto/federated_tables_service.rb b/app/services/carto/federated_tables_service.rb
index c8f00ce..322216a 100644
--- a/app/services/carto/federated_tables_service.rb
+++ b/app/services/carto/federated_tables_service.rb
@@ -62,13 +62,13 @@ module Carto

     def list_remote_schemas(federated_server_name, pagination)
       pagination[:offset] = (pagination[:page] - 1) * pagination[:per_page]
-      @user_db_connection[
+      @superuser_db_connection[
         select_remote_schemas_query(federated_server_name, pagination)
       ].all
     end

     def count_remote_schemas(federated_server_name)
-      @user_db_connection[
+      @superuser_db_connection[
         count_remote_schemas_query(federated_server_name)
       ].first[:count]
     end
```
I got:
```
ERROR:  permission denied for schema cartodb
LINE 1: SELECT cartodb.__CDB_FS_Generate_Server_Role_Name(server_int...
               ^
QUERY:  SELECT cartodb.__CDB_FS_Generate_Server_Role_Name(server_internal)
CONTEXT:  PL/pgSQL function __cdb_fs_list_foreign_schemas_pg(name) line 17 during statement block local variable initialization
PL/pgSQL function cdb_federated_server_list_remote_schemas(text) line 8 at RETURN QUERY
```

This PR is linked to `https://github.com/CartoDB/cartodb-postgresql/pull/382`. Update the local extension as usual.

cc/ @rafatower

Algunenano commented on spec/requests/carto/api/public/federated_tables_controller_spec.rb:
```diff
@@ -0,0 +1,718 @@
+      end
+    end
+
+    it 'returns 404 when there is not a faderated server with the provided name' do
```
"faderated" found twice :smile_cat:

dgaubert commented:
Same test is failing in https://github.com/CartoDB/cartodb/pull/15218

dgaubert commented:
retest this please

alrocar commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,285 @@
+        before_action :load_federated_server_attributes, only: [:register_federated_server, :update_federated_server ]
+        before_action :load_federated_server, only: [:update_federated_server, :unregister_federated_server, :show_federated_server]
+        before_action :check_federated_server, only: [:unregister_federated_server, :show_federated_server]
+        before_action :ensure_required_federated_server_attributes, only: [:register_federated_server, :update_federated_server]
```
Minor stuff, since callbacks are called in order from top to down, I'd ensure the required params is called before other checks.

dgaubert replied:
🆗

alrocar commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,285 @@
+
+        def check_permissions
+          @api_key = Carto::ApiKey.find_by_token(params["api_key"])
+          raise UnauthorizedError unless @api_key.master? || @api_key.dataset_metadata_permissions
```
`dataset_metadata_permissions` is already used by other API and OAuth scope. Do we want to use it for this other API? If that's the case maybe we can document it. See `datasets:metadata` scope in this [page](https://carto.com/developers/fundamentals/oauth-apps/)

dgaubert replied:
No, we don't want to use it yet.

alrocar commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,285 @@
+
+        def check_permissions
+          @api_key = Carto::ApiKey.find_by_token(params["api_key"])
+          raise UnauthorizedError unless @api_key.master? || @api_key.dataset_metadata_permissions
```
What happens if it does not find the api_key or I send an api_key of other user?

rafatower replied:
Ultimately permissions are enforced by DB and roles. The master api_key user's role is used to create a role to manage all FT's dependent objects. This API checks for DB errors, so I think it is pretty well covered overall.

This approach avoids duplication of ACL's in rails and in DB.

Please correct me if I'm wrong

alrocar replied:
Thanks @rafatower that's a good approach.

I was asking about the validation in this specific method, even when the permissions are enforce in the DB roles, if I send an invalid api_key we could raise an error directly from the API. A nice to have anyway :)

alrocar commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,362 @@
+
+    def register_federated_server_query(attributes)
+      config_server = {}
+      config_server["dbname"] = attributes[:dbname] unless (attributes[:dbname].nil? || attributes[:dbname] == "")
```
I guess you can write `attributes[:dbname].present?` instead

alrocar commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,362 @@
+      config_server["extensions"] = "postgis"
+      config_server["updatable"] = attributes[:mode] != 'read-write' ? "false" : "write"
+      config_server["use_remote_estimate"] = "true"
+      config_server["fetch_size"] = "1000"
```
this could be a constant

dgaubert replied:
I'm not able to initialize constant values with expressions... :(

alrocar replied:
I was referring to the "1000" but anyway, that's fine.

rafatower commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,285 @@
+        def list_remote_tables
+          result = @service.list_remote_tables(params[:federated_server_name], params[:remote_schema_name], @pagination)
+          # For unregistered tables we only want to keep the relevant properties
+          result.each {|table| table.slice!(:registered, :remote_schema_name, :remote_table_name, :columns) unless table[:registered]}
```
These look specially relevant to me, we're particularly interested in knowing the remote schema, table name and columns of a table that is not registered yet, or at least let the user decide, right? IMHO the main use case for this is to use that info to later register a foreign table

Algunenano replied:
I think you are reading this the opposite way, those are the only parameters to keep when the tables isn't registered.

dgaubert commented on NEWS.md:
```diff
@@ -38,6 +38,10 @@ sudo make install
+* Federated Tables:
+  * Add enpoints to list, register, update, and unregister Federated Servers
+  * Add enpoints to list Remote Schemas
+  * Add enpoints to list, register, update, and unregister Remote Tables
```
👌

dgaubert replied:
About the hyphens, is that a standard or just an internal backend rule?

alrocar replied:
we have a script `bump.rb` that converts the hyphens to * when we make a release of cartodb

rafatower commented on doc/developer-center/federated-tables-api/guides/01-quickstart.md:
```diff
@@ -0,0 +1,410 @@
+Param | Value | Description
+--- | --- | ---
+username | `documentation` | The CARTO account username.
+api_key | `bec1667cdedaa6fd70165f5099981d0c61ec1112` | The target CARTO account [API key](https://carto.com/developers/fundamentals/authorization/#api-keys).
```
I wouldn't grant this user/api_key the FT feature flag. I would state clearly it is used only for illustration purposes.

Algunenano replied:
It's a random api key, not a valid value. Although in truth is one or 2 characters off from the one in my local installation, but since that isn't exposed anywhere it's not an issue.

rafatower commented on doc/developer-center/federated-tables-api/reference/swagger.yaml:
```diff
@@ -0,0 +1 @@
+This should be a swagger file :D
```
Is this valid yaml/swagger? even if TBD we don't want doc generation to break because of it.

rafatower replied:
I'm missing a general description of what the whole thing is about (not really in the quickstart).

Apparently, and taking as an example the Import API, the place is right on this file:
https://carto.com/developers/import-api/
https://github.com/CartoDB/cartodb/blob/aeab65bb2d9bfbdeaa5a3403bd83a21da70519ab/doc/developer-center/import-api/reference/swagger.yaml#L7

As source of inspiration, there's this post: https://carto.com/blog/introducing-federated-tables/

Main concepts: 
- live queries
- sql and maps api
- self service api

rafatower commented on doc/developer-center/federated-tables-api/support/01-support-options.md:
```diff
@@ -0,0 +1,36 @@
+## Support Options
```
It is tempting to use a symlink instead of copy/paste (if/when we check it works)

dgaubert replied:
Sadly it was designed in that way, for instance, this document is duplicated in both `auth api` and `import api`. We are just following the current rules. The following is the `diff` between `doc/developer-center/import-api/support/01-support-options.md` and `doc/developer-center/auth-api/support/01-support-options.md`:

![Screenshot 2019-12-18 at 11 54 16](https://user-images.githubusercontent.com/1936888/71080288-24d0f200-218d-11ea-987e-f1d3171f8cd0.png)

In fact, they have some customization.

rafatower commented:
Forgot to mention: docs shall include a **beta banner**: "This feature is in **beta** stage and the API might change or have limited support"

dgaubert commented:
I did not miss that, I'm working on documentation https://github.com/CartoDB/developers/pull/675, the content of the documentation is kinda mixed in both projects. I'm understanding where is the better place to show a beta banner

gonzaloriestra commented on spec/services/carto/federated_tables_service_spec.rb:
```diff
@@ -0,0 +1,424 @@
+                federated_server_list = @service.list_servers(pagination)
+
+                expect(federated_server_list.length()).to eq(1)
+                expect(federated_server_list[0]).to have_key(:federated_server_name)
```
If you are already checking the value, there's no need to check if the key exists...

Algunenano commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,295 @@
+        end
+
+        def show_remote_table
+          render_jsonp(@remote_table, 200)
```
It's a pending bug: https://github.com/CartoDB/cartodb-platform/issues/6112

Algunenano replied:
It's a pending bug: CartoDB/cartodb-platform#6112 ( I deleted this message by accident)

dgaubert commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,294 @@
+
+        def get_error_message(exception)
+          regex = /^PG::(.*): ERROR:  /
+          message = exception.message.split("\n").find { |s| s.match(regex) }.to_s.gsub(regex, '')
```
I've tried it, but it doesn't return a thing.
```
[239, 248] in /vagrant/repos/cartodb/app/controllers/carto/api/public/federated_tables_controller.rb
   239:           render_jsonp(enriched_response, 200)
   240:         end
   241:
   242:         def rescue_from_service_error(exception)
   243:           debugger
=> 244:           message = get_error_message(exception)
   245:           case message
   246:           when /(.*) does not exist/
   247:             rescue_from_carto_error(Carto::LoadError.new(message))
   248:           when /Not enough permissions to access the server (.*)/
(byebug) exception.sql
nil
```

I'm going to skip this suggestion, thanks anyways! 

🤗

dgaubert commented on doc/developer-center/federated-tables-api/support/02-contribute.md:
```diff
@@ -0,0 +1,36 @@
+## Contribute
```
they are no the same document, for instance, a `diff` between `doc/developer-center/auth-api/support/02-contribute.md` and `doc/developer-center/import-api/support/02-contribute.md`:
![Screenshot 2019-12-18 at 11 21 28](https://user-images.githubusercontent.com/1936888/71077871-8b9fdc80-2188-11ea-9691-385706e1977e.png)

I wouldn't like to do anything different than we have in both import and auth documentation

dgaubert commented:
@gonzaloriestra, from https://github.com/CartoDB/cartodb/pull/15169#discussion_r358866840
> Also, is that the real master API key of that account? We shouldn't provide it.

No, it is not the real master API key of that account. In `doc/developer-center/import-api/guides/01-quickstart.md` we provide a fake API key as we are doing here.

dgaubert commented on doc/developer-center/federated-tables-api/guides/03-using-federated-tables.md:
```diff
@@ -0,0 +1,6 @@
+- SQL API
+- Tiler API
+
+Explain how caching works when a Federated Table is used in the different systems
```
🆗 
Sure, I'm going to open an issue for the next iterations to add this section.

dgaubert replied:
done https://github.com/CartoDB/cartodb/issues/15327

Algunenano commented on doc/developer-center/federated-tables-api/guides/02-general-concepts.md:
```diff
@@ -0,0 +1,17 @@
+**Warning:** Federated Tables is in **BETA** stage and the API might change or have limited support-
```
Extra `-` at the end instead of .

Algunenano replied:
The same sentence appears multiple times, with the same typo.

gonzaloriestra commented on doc/developer-center/federated-tables-api/guides/02-general-concepts.md:
```diff
@@ -0,0 +1,17 @@
+**Warning:** Federated Tables is in **BETA** stage and the API might change or have limited support-
```
Code suggestions are very handy for this kind of things :slightly_smiling_face: 
```suggestion
**Warning:** Federated Tables is in **BETA** stage and the API might change or have limited support.
```

Algunenano commented on doc/developer-center/federated-tables-api/guides/01-quickstart.md:
````diff
@@ -0,0 +1,413 @@
+#### Response
+
+The reponse has no extra content.
+```
````
Extra ```

dgaubert replied:
Delayed auto CRing! :trollface:

dgaubert commented on app/controllers/carto/api/public/federated_tables_controller.rb:
```diff
@@ -0,0 +1,295 @@
+          if message.present?
+            return message
+          else
+            raise exception.message
```
done

dgaubert commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,362 @@
+
+    # Federated Servers
+
+    def list_servers(pagination)
```
done

dgaubert commented on app/services/carto/federated_tables_service.rb:
```diff
@@ -0,0 +1,362 @@
+      unregister_table(
+        federated_server_name: attributes[:federated_server_name],
+        remote_schema_name: attributes[:remote_schema_name],
+        remote_table_name: attributes[:remote_table_name]
```
Now with keyword args, this comment is outdated, do you have a better approach?

gonzaloriestra commented:
Thanks for following so many suggestions :clap: 

One important question: why are you [updating the `lib/sql` submodule](https://github.com/CartoDB/cartodb/pull/15169/files#diff-3886c56375831f085f209bb0e50c8f22)? We should only do it when there is an upgrade in the extension and it requires more steps (update the versions in the code, add a notice in the NEWS...).

dgaubert commented:
> One important question: why are you updating the lib/sql submodule? We should only do it when there is an upgrade in the extension and it requires more steps (update the versions in the code, add a notice in the NEWS...).

Because we need the functions we implemented in https://github.com/CartoDB/cartodb-postgresql/releases/tag/0.33.0. And yes, you are right I've just updated the version extension and added a notice in NEWS.

Thanks for your help.
